### ISSUE TITLE
(research): climode model

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/somax#152]
# ClimODE Integration into SomaX

### Neural Advection as a Parametric Model Family

|                   |                                          |
|-------------------|------------------------------------------|
|**Repository**     |`jejjohnson/somax`                        |
|**Target module**  |`somax/_src/models/pde2d/`                |
|**Reference paper**|Verma et al., ICLR 2024 (arXiv:2404.10024)|
|**Stack**          |JAX · Equinox · Diffrax · FiniteVolX      |

-----

## 1. Motivation and Scope

ClimODE formulates weather forecasting as a continuous-time neural advection system grounded in the continuity equation from statistical mechanics. The model learns a flow velocity field that transports quantities conservatively, preventing the long-horizon forecast collapse that afflicts autoregressive jump-based approaches.

This document specifies how ClimODE is integrated into SomaX as a parametric model family — a new first-class resident of `somax._src.models.pde2d` alongside Burgers, Navier-Stokes, and linear convection. Three use cases within the broader library stack:

- **ML baseline** — compact differentiable model (~2–3M parameters) for benchmarking neural emulators on ERA5-style data
- **Differentiable prior** — physics-constrained prior for hybrid emulation and 4DVar/4DEnVar wiring through PipeKit + GaussX
- **Reanalysis prior** — the Gaussian emission head provides a likelihood model for data assimilation; the spatial observation operator H connects PipeKit to GaussX

> **Complexity ceiling.** ClimODE sits well below the SomaX hard ceiling of primitive equations. It is a 2D continuity-equation model with no hydrostatic assumption, no thermodynamic closure, and no vertical dimension — structurally equivalent to a multi-tracer shallow-water system with a learned velocity field.

The scope is strictly **integration, not new filtering algorithms**. GaussX already exceeds reference implementations on DA math; the work here is wiring ClimODE’s emission model into GaussX’s likelihood interface.

-----

### 2.1 Governing Equations

ClimODE models a spatiotemporal process `u(x,t) ∈ R^(K×H×W)` of K quantities over a 2D domain. The dynamics are a second-order system in `(u, v)`:

```
∂u_k/∂t  =  −∇·(u_k v_k)                          # conservative continuity
∂v_k/∂t  =  f_θ(u, ∇u, v, ψ(x,t))                # neural flow acceleration
```

The first equation is value-preserving by construction: `∫u_k(x,t) dx = const` for all t, k. This global conservation prevents long-horizon forecast collapse without any explicit regularisation. The second equation drives the transport velocity via a neural network `f_θ`; `v` is a prognostic variable, not diagnosed from `u`.

### 2.2 Method of Lines Reduction

Applying MOL on a spatial grid reduces the PDE to a first-order ODE system in the joint state `(u, v)`:

```
d/dt [u(t)]   =   [−∇·(u_k v_k)]_k
     [v(t)]       [f_θ(u, ∇u, v, ψ)]_k
```

This plugs directly into `diffrax.diffeqsolve` via `ODETerm`. No custom integrators required.

### 2.3 Spatiotemporal Embedding

```
ψ(t)    = [sin(2πt), cos(2πt), sin(2πt/365), cos(2πt/365)]         # 4 channels
ψ(x)    = [{sin,cos}×{h,w}, sin(h)cos(w), sin(h)sin(w)]             # 6 channels
ψ(x,t)  = [ψ(t), ψ(x), ψ(t)⊗ψ(x), ψ(h), ψ(w), lsm, oro]          # C channels
```

The static part `ψ(x)` is precomputed once and stored as a `static=True` field. The time-dependent `ψ(t)` is evaluated inside `vector_field` from the scalar `t` argument.

### 2.4 Initial Velocity Estimation

The second-order system requires `v̂(t₀)`. Using the identity `∂u/∂t + ∇·(uv) = 0`, this is a penalised least-squares solve:

```
v̂(t₀) = argmin_v  ‖u̇(t₀) + ∇·(u(t₀)·v)‖²  +  α‖v‖²_K
```

where `u̇` is a finite-difference estimate from past states and the Tikhonov regulariser with RBF kernel K enforces spatial smoothness. This is a **standalone utility function**, not a model method — it runs once before integration.

### 2.5 Emission Model

A learned Gaussian emission head handles the open-system correction and predictive uncertainty:

```
μ_k(x,t), σ²_k(x,t)  =  g_k(u(x,t), ψ(x,t))
y^obs_k(x,t)           ~  N(u_k + μ_k,  σ²_k)
```

The bias `μ` models residual sources/sinks (e.g. diurnal cycle); `σ²` gives aleatoric + epistemic uncertainty. The emission model is **not part of the dynamics** — it lives at the PipeKit/GaussX boundary as an observation likelihood.

-----

### 3.1 Module Layout

ClimODE lives in `somax._src.models.pde2d` alongside the existing scalar-field PDE models — not in `swm/` or `qg/`, because it carries no GFD assumptions.

```
somax/_src/models/pde2d/
    climode.py          # ClimODEState, ClimODEParams, ClimODE
    climode_nets.py     # f_theta architectures: ConvNet, HybridNet
```

### 3.2 State and Parameter Types

```python
class ClimODEState(State):
    u: Float[Array, "K Ny Nx"]    # K quantities at T-points
    v: Float[Array, "2K Ny Nx"]   # per-quantity: (vx_k, vy_k) for each k
                                   # OR "2 Ny Nx" if shared_velocity=True

class ClimODEParams(Params):
    flow_net: eqx.Module           # f_theta — any equinox module
                                   # all weights are JAX leaves → jax.grad
```

> **Per-quantity vs. shared velocity.** Per-quantity (paper default): `v` is `(2K, Ny, Nx)` — each quantity has its own transport velocity. More expressive, but 5× the velocity parameters for K=5. Shared: `v` is `(2, Ny, Nx)`, one wind field advects all tracers. Physically motivated, cheaper, and the better SomaX default. Controlled by `shared_velocity: bool = eqx.field(static=True, default=True)`.

### 3.3 Core Model: `ClimODE`

```python
class ClimODE(SomaxModel):
    params:           ClimODEParams
    grid:             CartesianGrid2D    # static
    diff:             Difference2D
    interp:           Interpolation2D
    mask:             Mask2D | None
    spatial_embed:    Float[Array, "C Ny Nx"]   # static: ψ(x)
    K:                int                # static
    shared_velocity:  bool               # static
```

#### 3.3.1 Conservative Continuity Term

Reuses FiniteVolX’s existing flux-divergence operators — the same operators used in `LinearConvection2D`. No new numerics required.

```python
def _continuity_term(self, u_k, vx_k, vy_k):
    flux_x = self.interp.T_to_U(u_k) * self.interp.T_to_U(vx_k)
    flux_y = self.interp.T_to_V(u_k) * self.interp.T_to_V(vy_k)
    return -(
        self.diff.diff_x_U_to_T(flux_x)
      + self.diff.diff_y_V_to_T(flux_y)
    )
```

This gives discrete mass conservation on the C-grid to machine precision — the same guarantee as the SWM mass equation.

#### 3.3.2 Neural Flow Acceleration

```python
# Input tensor: (5K + C + 4, Ny, Nx)
grad_u   = self._spatial_gradients(u)        # (2K, Ny, Nx)
psi_t    = _time_embed(t)                    # (4,) → broadcast
psi_xt   = concat([spatial_embed, psi_t])    # (C+4, Ny, Nx)
net_in   = concat([u, grad_u, v, psi_xt])    # (5K+C+4, Ny, Nx)
dv_dt    = self.params.flow_net(net_in)      # (2K or 2, Ny, Nx)
```

#### 3.3.3 Integration

The base `SomaxModel.build_terms()` wraps `vector_field` in a single `dfx.ODETerm` — no override needed. The joint state `(u, v)` is already first-order.

```python
state0   = ClimODEState(u=u0, v=estimate_initial_velocity(...))
state_6h = model.step(state0, dt=6*3600.0)
sol      = model.integrate(state0, t0=0., t1=36*3600.,
                           dt=1800., saveat=dfx.SaveAt(ts=...))
```

### 3.4 Network Architectures (`climode_nets.py`)

Both produce `(2K, Ny, Nx)` or `(2, Ny, Nx)` depending on velocity mode. The model class is architecture-agnostic — any `eqx.Module` with the right shape works.

**`ClimODEConvNet` (local effects)**

ResNet-style 3×3 convolutional blocks. Channel-first format `(C, H, W)` matches `(K, Ny, Nx)` throughout — no transposes needed.

```python
class ClimODEConvNet(eqx.Module):
    # Input:  (5K+C+4, Ny, Nx)
    # Output: (2K or 2, Ny, Nx)
    layers: list[eqx.nn.Conv2d]
```

**`ClimODEHybridNet` (local + global effects)**

Adds a KQV attention branch to capture long-range teleconnections. `gamma` is a differentiable scalar leaf — fully visible to `jax.grad`.

```python
class ClimODEHybridNet(eqx.Module):
    conv_net:  ClimODEConvNet
    attn_net:  ClimODEAttentionNet   # KQV with CNN projections
    gamma:     Array                 # learnable blend weight

    def __call__(self, x):
        return self.conv_net(x) + self.gamma * self.attn_net(x)
```

### 3.5 Emission Model

Lives outside the dynamics layer, at the PipeKit/GaussX seam:

```python
class ClimODEEmission(eqx.Module):
    net: eqx.Module   # (K+C, Ny, Nx) → (2K, Ny, Nx): [mu, log_sigma]

    def log_likelihood(self, y_obs, u, psi_xt):
        mu, log_sigma = jnp.split(self.net(concat([u, psi_xt])), 2, axis=0)
        return jax.scipy.stats.norm.logpdf(
            y_obs, u + mu, jnp.exp(log_sigma)).sum()
```

The observation operator H (PipeKit → GaussX bridge) maps the emission’s `μ/σ` output to the assimilation likelihood. Integration plumbing only — no new DA math.

-----

### 4.1 What Already Exists

|Component                         |Used for                                                        |Status          |
|----------------------------------|----------------------------------------------------------------|----------------|
|`SomaxModel`                      |`vector_field` + BC contract, `build_terms`, `step`, `integrate`|`core/model.py` |
|`dfx.ODETerm` / `diffeqsolve`     |ODE integration of joint `(u,v)` state                          |core diffrax dep|
|`CartesianGrid2D` + `Difference2D`|`∇·(uv)` via `diff_x_U_to_T`, `diff_y_V_to_T`                   |FiniteVolX      |
|`Interpolation2D`                 |T→U, T→V face interpolation for flux construction               |FiniteVolX      |
|`eqx.Module` + `eqx.nn.Conv2d`    |Neural network `f_θ` as a differentiable pytree                 |Equinox         |
|`State` / `Params` / `Diagnostics`|Type hierarchy for ClimODE state and params                     |`core/types.py` |
|`jax.scipy.stats.norm`            |Emission log-likelihood                                         |JAX stdlib      |
|GaussX covariance + EnKF          |DA wiring for emission model                                    |GaussX          |

### 4.2 What Is New

- `climode.py` — joint `(u,v)` state, conservative continuity term, spatiotemporal embedding, flow net dispatch
- `climode_nets.py` — `ClimODEConvNet` and `ClimODEHybridNet`
- `estimate_initial_velocity()` — standalone Tikhonov least-squares utility for `v̂(t₀)`
- `ClimODEEmission` — Gaussian emission head and log-likelihood interface

### 4.3 Cross-Cutting Relaxation Primitive

The emission bias `μ` is structurally identical to the relaxation/nudging pattern `coef·W·(X_ref − X)` identified as a recurring cross-cutting primitive across FiniteVolX, xrtoolz, and GaussX. The emission model is that primitive with a learned rather than prescribed coefficient.

### 4.4 PipeKit Integration

`ClimODE` satisfies the `pipekit_cycle.ForwardModel` protocol structurally via `SomaxModel.step()` and `state_signature` — no pipekit import in the model itself.

```python
# PipeKit cycle sketch — no new primitives needed
cycle = Cycle(
    forward      = ClimODE.create(...),       # step() interface
    obs_operator = H,                         # spatial obs operator
    likelihood   = ClimODEEmission(...),      # GaussX bridge
    filter       = gaussx.EnKF(...),          # existing GaussX
)
```

-----

## 5. Key Design Decisions

|Decision                 |Options                                            |Recommendation                             |Rationale                                                                                                       |
|-------------------------|---------------------------------------------------|-------------------------------------------|----------------------------------------------------------------------------------------------------------------|
|**Velocity mode**        |Per-quantity `(2K,Ny,Nx)` vs. shared `(2,Ny,Nx)`   |Shared default; per-quantity as static flag|Shared is physically motivated (all tracers in one wind), cheaper (2 vs 2K channels for K=5), and a better prior|
|**IMEX splitting**       |Single `ODETerm` vs. `MultiTerm` IMEX              |Single `ODETerm` to start                  |Neither the continuity term nor NN evaluation is stiff in the parabolic sense; add IMEX later if needed         |
|**Initial velocity**     |Tikhonov least-squares vs. encoder NN              |Tikhonov utility function                  |Avoids a second network to train; differentiable through `jax.scipy.optimize` if 4DVar through `v₀` is needed   |
|**Emission placement**   |Inside `SomaxModel` vs. separate module            |Separate `ClimODEEmission`                 |Keeps dynamics layer clean and reusable; emission is an observation model, not dynamics                         |
|**Architecture coupling**|Bake `ConvNet` into `ClimODE` vs. `eqx.Module` slot|`eqx.Module` slot                          |Any equinox module with the right shape works; swap architectures without touching the model class              |

-----

### 6.1 Five-PR Rollout

|PR|Content                                                                                                         |Verification                                                                                          |
|--|----------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------|
|1 |`climode.py`: `ClimODEState`, `ClimODEParams`, `ClimODE` with identity flow_net and conservative continuity term|`∫u dx = const` over 10 steps with uniform IC, to machine precision                                   |
|2 |`climode_nets.py`: `ClimODEConvNet` — 3-layer ResNet with 3×3 `Conv2d`                                          |Output shape `(2K,Ny,Nx)`; param count 2–3M; `jax.grad` non-zero on all weights                       |
|3 |`climode_nets.py`: `ClimODEHybridNet` — attention branch + learnable `gamma`                                    |`jax.grad` w.r.t. `gamma` non-zero; attention output contributes to `dv/dt`                           |
|4 |`estimate_initial_velocity()` + spatiotemporal embedding `ψ(x,t)`                                               |`v̂` reduces `‖u̇ + ∇·(uv̂)‖₂` vs `v=0` on an ERA5 snapshot; `ψ` shape matches expected channels         |
|5 |`ClimODEEmission` + `__init__.py` exports + full test suite                                                     |Log-likelihood shapes correct; `jax.grad` through emission w.r.t. NN params; GaussX interop smoke test|

### 6.2 Verification Ladder

1. **1D Gaussian blob** — advection under prescribed constant velocity. Exact solution known; tests `∫u dx` conservation and spatial order.
1. **2D rotating Gaussian** — solid-body rotation under prescribed `(vx,vy) = (−y,x)`. Tests 2D flux divergence on the C-grid.
1. **Neural velocity self-consistency** — train on a double-gyre snapshot; confirm training loss (eq. 12 in paper) decreases and emission variance `σ²` is bounded.
1. **ERA5 WeatherBench subset** — 5 variables at 5.625°, 10-year training split. Not a CI gate but the stop-condition for architecture tuning.

### 6.3 Items to Verify Before Coding

- Whether `Difference2D` exposes a cell-centred gradient (T→T) or only face-centred differences (T→U, T→V). The spatial gradient `∇u` fed to the network is at T-points; the round-trip T→U→T suffices.
- Exact channel count C for the spatiotemporal embedding — depends on whether `lsm`/`oro` are included as static constants or left as optional args.
- Whether `jax.vmap` over the K-quantity axis is the right approach for the continuity term, or whether a batch axis on `Difference2D` is more efficient.

-----

### 7.1 Near-Term

- **Positivity preservation** — the conservative flux form can produce negative values for quantities that must stay non-negative (e.g. humidity). Should a flux limiter (monotone reconstruction) be applied in the continuity term, consistent with the FiniteVolX positivity principle?
- **4DVar through `v₀`** — if the initial velocity is estimated via Tikhonov, is `jax.scipy.optimize.minimize` differentiable enough for 4DVar gradient flow, or should `v₀` be a free parameter in the assimilation state?
- **Sphere vs. plane** — the paper operates on a lat-lon grid; SomaX currently uses `CartesianGrid2D`. Lat-lon metric corrections (cosine weighting) are deferred but should be flagged.

### 7.2 Longer Horizon

- **SpectralDiffX backend** — spatial gradients and divergence could use SpectralDiffX instead of FiniteVolX for spectral accuracy on periodic domains. Natural follow-on after the FiniteVolX baseline.
- **3D extension** — adding a vertical dimension (pressure levels) would push toward the primitive-equation ceiling. Architecturally straightforward; add a third spatial dim to state and operators.
- **NeuralGCM hybrid** — the ClimODE velocity field is structurally similar to NeuralGCM’s parameterized tendency. The `ConvNet`/`HybridNet` architectures here could serve as the parameterization network in a NeuralGCM-style hybrid, following the staged NeuralGCM recommendation (semi-implicit stepper, spectral transforms, Held-Suarez benchmark).
- **Ensemble emission** — replace the single Gaussian emission with an ensemble sampler using GaussX’s existing EnKF infrastructure, giving Monte Carlo uncertainty quantification for free.

-----

## 8. References

- Verma, Y., Heinonen, M., & Garg, V. (2024). *ClimODE: Climate and Weather Forecasting with Physics-Informed Neural ODEs.* ICLR 2024. arXiv:2404.10024.
- Johnson, J. E. (2024). *somax: Simple Ocean Models with JAX.* github.com/jejjohnson/somax.
- Johnson, J. E. (2024). *finitevolX: Finite Volume Methods on Arakawa C-Grid.* github.com/jejjohnson/finitevolX.
- Kidger, P. (2021). *On Neural Differential Equations.* PhD thesis, University of Oxford. (Diffrax library.)
- Bradbury, J. et al. (2018). *JAX: Composable Transformations of Python+NumPy Programs.* github.com/google/jax.
